### PR TITLE
Python: Plan Object Model

### LINKED ISSUE #1486: Python: Implement POM Plan Object Model
Planner has been updated to be a new object model instead of a skill:
https://github.com/microsoft/semantic-kernel/pull/255

### PR DESCRIPTION
**Pull request by awharrison-28**
### Motivation and Context
This PR ports the plan object model found in the .NET SK to Python. This fundamentally changes the concept of a planner from a `semantic skill` to a unique type of `SKFunction`. With this model, we can begin implementing sequential planner and offer much better plan customization. 

**Notes:**
- This PR does not implement plan serialization/deserialization to/from JSON. There is a huge existing effort to make the Python SK completely Pydantic enabled, which will bring serialization and deserialization capability with it.
- This PR does not include ports of `Action Planner` and `Sequential Planner` to the Plan Object Model (POM). Following PRs will address this.
- I also added some missing copywrite headers. A future PR will add headers to any still missing them. 
- added a VSCode launch configuration to run the current file with pytest
- closes #1486 

### Description
- Added file called `plan.py` which is a python port of the Plan.cs file found in the .NET SK. Note that it does not contain serialization logic at this time.
- Added new kernel error `FunctionInvokeError` - mirrors the error list of the .NET SK
- Added tests for Plan creation
- Added tests for Plan execution - both sync and async

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [ ] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: python/semantic_kernel/planning/plan.py [hunk 1]
```diff
@@ -1,17 +1,432 @@
...
+                skill_collection=context.skills,
+                logger=context.log,
+            )
+            result = await step.invoke_async(context=func_context)
+            result_value = result.result
+
+            if result.error_occurred:
+                raise KernelException(
+                    KernelException.ErrorCodes.FunctionInvokeError,
+                    "Error occured while running plan step: "
```
Comment (shawncal) [inline python/semantic_kernel/planning/plan.py:301]:
```suggestion
                    "Error occurred while running plan step: "
```

**Comment by alexchaomander** [comment]
@joowon-dm-snu @Sailesh-2209 As mentioned in #1770, please take a look at the Plan Object Model and leave any comments before implementing future planners!

**Review by alexchaomander — DISMISSED** [review]
Looks good to me!

File: .vscode/launch.json [hunk 2]
```diff
@@ -6,34 +6,35 @@
       "type": "coreclr",
       "request": "launch",
       "preLaunchTask": "build (CopilotChatWebApi)",
-      // If you have changed target frameworks, make sure to update the program path.
```
Comment (shawncal) [inline .vscode/launch.json:9]: Seem unrelated. Revert these comment removals?

File: python/semantic_kernel/core_skills/math_skill.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 # Copyright (c) Microsoft. All rights reserved.
 
+
```
Comment (shawncal) [inline python/semantic_kernel/core_skills/math_skill.py:3]: Can we align to one line of whitespace?

Comment (shawncal) [inline python/semantic_kernel/core_skills/math_skill.py:3]:
```suggestion
```

File: python/semantic_kernel/core_skills/text_skill.py [hunk 4]
```diff
@@ -1,3 +1,6 @@
+# Copyright (c) Microsoft. All rights reserved.
+
+
```
Comment (shawncal) [inline python/semantic_kernel/core_skills/text_skill.py:3]:
```suggestion
# Copyright (c) Microsoft. All rights reserved.

```

File: python/semantic_kernel/planning/plan.py [hunk 5]
```diff
@@ -1,17 +1,432 @@
...
+            # loop through steps until completion
+            while self.has_next_step:
+                function_context = context
+                self.add_variables_to_context(self._state, function_context)
+                await self.invoke_next_step(function_context)
+                self.update_context_with_outputs(context)
+
+        return context
+
+    def invoke(
```
Comment (shawncal) [inline python/semantic_kernel/planning/plan.py:157]: general question, can `invoke` call and wrap `invoke_async`, or is that considered bad practice? I'm just a little worried about having dual implementations of all the context variable handling, in case they get out of alignment. Good opportunity for some refactoring. (future PR++)

Comment (awharrison-28) [reply python/semantic_kernel/planning/plan.py:157]: I noticed it to. I think we should do this yes (in a future PR)

File: .vscode/launch.json (hunk 2 above)
Comment (awharrison-28) [reply .vscode/launch.json]: i think the intellisense did it, added comments back

**Review by alexchaomander — APPROVED** [review]
